Plugin demos (#13)
Cr1t-GYM/Jenkinsfile-runner-actions-demo: demo/s3-artifact-manager/README.md

# How to copy the pipeline results to an external object storage

This example shows how to copy the pipeline results to an external object storage such as `AWS S3` by using
[artifact-manager-s3](https://plugins.jenkins.io/artifact-manager-s3/). In this example, it takes `AWS S3` as an example
and other object storage services should work as well.

If you use `AWS S3` like this example, you need to know your the region of the user, `AWS access key id`, `AWS secret access key`, 
`AWS S3 bucket name` and `AWS S3 bucket prefix` in advance. For example, if your `S3 URI` for a specific folder under a 
bucket is `s3://jenkins-action/test/`, `AWS S3 bucket name` is `jenkins-action` and `AWS S3 bucket prefix` is `test/`.
`AWS access key id` and `AWS secret access key` can be downloaded when you create a new user for your account.
Furthermore, in order to make this AWS access key have the ability to manage your S3 storage,

> you need to give it `AmazonS3FullAccess` in the IAM user permission page.

Reviewer: ```suggestion
you will need to grant it `AmazonS3FullAccess` in the IAM user permission page.
```

> Here is the step-by-step example in `jfr-container-action`.

Reviewer: ```suggestion
Here is the step-by-step example for `jfr-container-action`:
```

You can check the documentations for [archiveArtifacts](https://www.jenkins.io/doc/pipeline/steps/core/#archiveartifacts-archive-the-artifacts),
[unarchive](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#unarchive-copy-archived-artifacts-into-the-workspace),
[stash](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#stash-stash-some-files-to-be-used-later-in-the-build) and 
[unstash](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#unstash-restore-files-previously-stashed).
In this example, it only shows how to use `archiveArtifacts` for uploading your pipeline results.
It assumes there is a maven project in your workspace and `artifact-manager-s3` will help you upload the compiled targets to the `AWS S3` after
the pipeline builds the sources successfully.
```groovy
pipeline {
    agent any
    tools {
        maven 'maven'
    }
    stages {
        stage('build') {
            steps {
                sh 'mvn clean install -B --no-transfer-progress'
            }
        }
    }
    post {
        success {
            archiveArtifacts 'target/*.jar'
        }
    }
}
```

> 5. In your workflow definition, you need to set up your credentials and `jfr-container-action` correctly.

Reviewer: ```suggestion
5. In your workflow definition, you will need to set up your credentials and `jfr-container-action` correctly.
```
